File: basic/src/my_py_pkg/my_py_pkg/minimal_publisher.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from example_interfaces.msg import String


class MinimalPublisher(Node):
    """Publisher example that periodically sends out a string"""

    def __init__(self):
        """Constructor"""

        # call the Node class constructor with the node name
        super().__init__("minimal_publisher")

        # create a publisher object
        self._publisher = self.create_publisher(String, 'my_topic', 10)
        
        # Periodically call method
        self._timer = self.create_timer(0.5, self._timer_callback)
        
        # Counter for message send
        self._counter = 0
        
    def _timer_callback(self):
        """Publishes a simple message to topic"""

        # Fill out string message
        msg = String()
        msg.data = f"Hello {self._counter}"

        # publish message to topic
        self._publisher.publish(msg)
        self.get_logger().info(f"Publisher: {msg.data}")

        # Increment counter
        self._counter += 1
            

def main(args=None):
    """Main entrypoint"""

    # Initialize and run node
    try:
        rclpy.init()
        node = MinimalPublisher()
        rclpy.spin(node)

    # Catch ctrl+c or shutdown request
    except(KeyboardInterrupt, ExternalShutdownException):
        ...
        
    finally:
        if node is not None:
            node.destroy_node()
            
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main() 

File: basic/src/my_cpp_pkg/src/minimal_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

/*
Publisher example hat periodically sends out a string
*/
class MinimalPublisher : public rclcpp::Node
{
public:
    /*
    Constructor (call Node class constructor with node name)
    */

    MinimalPublisher() : Node("minimal_publisher")
    {
        // create a publisher object
        publisher_ = this->create_publisher<example_interfaces::msg::String>(
            "my_topic",
            10);

        // Periodically call method
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(500),
            std::bind(&MinimalPublisher::timer_callback, this));

        // Counter for message send
        counter_ = 0;
    }

private:
    void timer_callback()
    {
        auto msg = example_interfaces::msg::String();
        msg.data = "Hello World " + std::to_string(counter_);

        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Publishing: %s", msg.data.c_str());

        counter_++;
    }
    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    size_t counter_;
};

int main(int argc, char *argv[])
{
    // Initialize ros2
    rclcpp::init(argc, argv);

    // Initialize and run node
    auto node = std::make_shared<MinimalPublisher>();
    rclcpp::spin(node);

    // cleanup
    rclcpp::shutdown();

    return 0;
}